module/os/linux/zfs/zfs_vfsops.c

```diff
@@ -1423,6 +1426,9 @@ zfsvfs_teardown(zfsvfs_t *zfsvfs, boolean_t unmounting)
 		txg_wait_synced(dmu_objset_pool(zfsvfs->z_os), 0);
 	}
 	dmu_objset_evict_dbufs(zfsvfs->z_os);
+	dsl_dir_t *dd = os->os_dsl_dataset->ds_dir;
+	dd->dd_activity_cancelled = B_TRUE;
+	cv_broadcast(&dd->dd_activity_cv);
```

Comment:
Why aren't we using dsl_dir_cancel_waiters() here, which waits for the waiters to release their holds?